src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataResourceDeserializer.cs

```diff
@@ -358,6 +422,197 @@ public virtual object CreateResourceInstance(IEdmStructuredTypeReference structu
+        { 
+            Contract.Assert(readContext != null);
+
+            if (resourceWrapper?.ResourceBase?.Id == null)
+            {
+                return resourceWrapper;
+            }
+
+            IEnumerable<ODataProperty> keys = CreateKeyProperties(resourceWrapper.ResourceBase.Id, readContext);
+            if (keys == null)
+            {
+                return resourceWrapper;
+            }
+
+            if (resourceWrapper.ResourceBase.Properties == null)
+            {
+                resourceWrapper.ResourceBase.Properties = keys;
+            }
+            else
+            {
+                IDictionary<string, ODataProperty> newPropertiesDic = resourceWrapper.ResourceBase.Properties.ToDictionary(p => p.Name, p => p);
+                foreach (ODataProperty key in keys)
+                {
+                    if (!newPropertiesDic.ContainsKey(key.Name))
+                    {
+                        newPropertiesDic[key.Name] = key;
+                    }
+                }
+
+                resourceWrapper.ResourceBase.Properties = newPropertiesDic.Values;
+            }
+
+            return resourceWrapper;
+        }
+
+        /// <summary>
+        /// Do uri parsing to get the key values.
+        /// </summary>
+        /// <param name="id">The key Id.</param>
+        /// <param name="readContext">The reader context.</param>
+        /// <returns>The key properties.</returns>
+        private static IList<ODataProperty> CreateKeyProperties(Uri id, ODataDeserializerContext readContext)
+        {
+            Contract.Assert(id != null);
+            Contract.Assert(readContext != null);
+            IList<ODataProperty> properties = new List<ODataProperty>();
+            if (readContext.Request == null)
+            {
+                return properties;
+            }
+           
+            ODataPath odataPath = GetODataPath(id.OriginalString, readContext);
+            if (odataPath?.Segments != null)
+            {
+                KeySegment keySegment = odataPath.Segments.OfType<KeySegment>().LastOrDefault();
+
+                if (keySegment != null)
+                {
+                    foreach (KeyValuePair<string, object> key in keySegment.Keys)
+                    {
+                        properties.Add(new ODataProperty
+                        {
+                            Name = key.Key,
+                            Value = key.Value
+                        });
+                    }
+                }
+            }
+
+            return properties;                     
+        }
+
+        private static ODataPath GetODataPath(string id, ODataDeserializerContext readContext)
+        {
+            try
+            {
+                IODataPathHandler pathHandler = readContext.InternalRequest.PathHandler;
+                IWebApiRequestMessage internalRequest = readContext.InternalRequest;
```

Comment:
Nit: you can simplify slightly:

                IWebApiRequestMessage internalRequest = readContext.InternalRequest;
                IODataPathHandler pathHandler = internalRequest.PathHandler;